python/ql/src/experimental/semmle/python/frameworks/Log.qll

```diff
@@ -0,0 +1,92 @@
+/**
+ * Provides classes modeling security-relevant aspects of the log libraries.
+ */
+
+private import python
+private import semmle.python.dataflow.new.DataFlow
+private import semmle.python.dataflow.new.TaintTracking
+private import semmle.python.dataflow.new.RemoteFlowSources
+private import experimental.semmle.python.Concepts
+private import semmle.python.frameworks.Flask
+private import semmle.python.ApiGraphs
+
+/**
+ * Provides models for Python's log-related libraries.
+ */
+private module log {
+  /**
+   * Log output method list.
+   *
+   * See https://docs.python.org/3/library/logging.html#logger-objects
+   */
+  private class LogOutputMethods extends string {
+    LogOutputMethods() { this in ["info", "error", "warn", "warning", "debug", "critical"] }
+  }
+
+  /**
+   * The class used to find the log output method of the `logging` module.
+   *
+   * See `LogOutputMethods`
+   */
+  private class LoggingCall extends DataFlow::CallCfgNode, LogOutput::Range {
+    LoggingCall() {
+      this = API::moduleImport("logging").getMember(any(LogOutputMethods m)).getACall()
+    }
+
+    override DataFlow::Node getAnInput() { result = this.getArg(_) }
```

Comment:
Note that `getArg(_)`, perhaps confusingly, will not return all arguments. In particular, starred arguments are filtered out, as are keyword only arguments. This means that for `info(msg, *args, **kwargs)` you will see neither `args`, nor `kwargs`.

Note also, that if you add `log`, its first argument is `level` which should probably not be considered an input.

Reply:
Pr updated.